File: 0113-path-sum-ii/0113-path-sum-ii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function(root, targetSum) {
    return findPath(root, targetSum, [], []);
};

function findPath(root, sum, solution, result) {
    if (!root) return result;
    
    solution.push(root.val);
    if (!root.left && !root.right && root.val == sum) {
        result.push(solution.slice());
    }
    
    findPath(root.left, sum - root.val, solution, result);
    findPath(root.right, sum - root.val, solution, result);
    
    solution.pop();
    
    return result;
}

File: 0113-path-sum-ii/0113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
      
      def path_finder(root):
        if not root:
          return []
        if not root.left and not root.right:
          return [[root.val]]

        all_paths = []
        left_paths = path_finder(root.left)
        for path in left_paths:
          all_paths.append([ root.val, *path ])

        right_paths = path_finder(root.right)
        for path in right_paths:
          all_paths.append([ root.val, *path ])
          
        return all_paths
      
      paths = path_finder(root)
      result = []
      for sub_path in paths:
        if sum(sub_path) == targetSum:
          result.append(sub_path)
          
      return result
    
        